<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="ShoRAH : Short Reads Assembly into Haplotypes" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">
    <link rel="shortcut icon" />
    <title>Shorah</title>
  </head>

  <body>
    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/ozagordi/shorah">View on GitHub</a>

          <h1 id="project_title" href="/index.html">ShoRAH</h1>
          <h2 id="project_tagline">Short Reads Assembly into Haplotypes</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/ozagordi/shorah/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/ozagordi/shorah/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>


    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">

    <h3 id='generally_speaking'>Generally speaking</h3>

<p>Two kinds of local analysis are possible, one invoked by the program <code>dec.py</code>, the other by <code>amplian.py</code>. We will define these two as shotgun mode (albeit improperly) and amplicon mode. The input is always a <a href='input.html'>sorted bam file</a>.</p>

<h4 id='shotgun_mode'>Shotgun mode</h4>

<p>This is performed when the amplicon is longer than the typical read length and we don&#8217;t want to focus our analysis on a specific region of this amplicon. <code>dec.py</code> cuts on the aligment a set of overlapping regions (parameters can be adjusted, run <code>dec.py -h</code> for help), and passes them to <code>diri_sampler</code> for error correction and local haplotype reconstruction. This is the first step of the <a href='global.html'>global analysis</a>. Additionally, it runs <code>snv.py</code> for SNV calling.</p>

<h4 id='amplicon_mode'>Amplicon mode</h4>

<p>When the amplicon size is smaller than the read length or comparable to it, it is in general more appropriate to use <code>amplian.py</code>, which extracts the multiple alignment corresponding to the specified amplicon. By default, this includes reads that cover at least 95% of the amplicon region (this can be adjusted with the option <code>-m MIN_OVERLAP</code>). It then runs <code>diri_sampler</code> for error correction and <code>snv.py</code> for SNV calling.</p>

<h3 id='examples'>Examples</h3>

<h4 id='shotgun_mode_'>Shotgun mode: <code>dec.py</code></h4>

<p>Run <code>dec.py -h</code> for help.</p>

<pre><code>[user@host]$ dec.py -h
Usage: dec.py [options]
Options:
-h, --help   show this help message and exit
-b IN_BAM, --bam=IN_BAM   file with aligned reads in .bam format
-f IN_FASTA, --fasta=IN_FASTA   reference genome in fasta format
. . . more options . . .</code></pre>

<p>You should specify the window size (default is 201), add the <code>-k</code> flag if you want to save intermediate files (for example, to check the convergence of the Markov chain). Next line runs with a window length of 300 and keeps intermediate files. <code>ref_file.fasta</code> contains the reference sequence that has been used to align reads.</p>

<pre><code>[user@host]$ dec.py -b my_sorted_bam.bam -f ref_file.fasta -w 300 -k</code></pre>

<h4 id='amplicon_mode_'>Amplicon mode: <code>amplian.py</code></h4>

<p>As usual, the help is available with <code>-h</code>.</p>

<pre><code>[user@host]$ amplian.py -b my_sorted_bam.bam -f amplicon.fasta -m 0.95</code></pre>

<p><code>amplian.py</code> uses the option <code>-m</code> to specify minimum overlap that the reads are required to have with the amplicon in order to be included (default 95%). The file <code>amplicon.fasta</code> contains the amplicon sequence used to align reads, which should be slightly smaller than the typical read length.</p>

<h3 id='output_of_local_analysis'>Output of local analysis</h3>

<h4 id='shotgun_mode'>Shotgun mode</h4>

<p>Suppose we are analyzing a region of 1500 bp, and we chose window size of 300. By default, windows overlap such that (except for the boundaries of the region) each base is covered by three windows (option <code>-s</code>). So <code>dec.py</code> will produce windows 1-300, 101-400, 201-500, 301-600 and so on. All these are passed to <code>diri_sampler</code> and local haplotype are inferred. For each window there will be several files produced. We describe here two of these, the <strong>support</strong> and the <strong>sampling</strong> file.</p>

<p>The support file has the <code>support.fas</code> suffix. It is a fasta file with the reconstructed haplotypes. The header of each sequence contains information on the quality of the reconstruction (posterior probability of the haplotype) and the frequency of the haplotype. For example,</p>

<pre><code>&gt;hap_0|posterior=1 ave_reads=235.168
CCTCAGATCACTCTTTGGCAACGACCCTTCGTC</code></pre>

<p>indicates that the reconstructed haplotype has a posterior probability of 1 (it was always present in the sampled Markov chain) and on average 235 reads were assigned to it. Good results have been obtained by discarding haplotypes with less than 90-95% support.</p>

<p><code>snv.py</code> takes care of parsing the single nucleotide variants, check the strand bias and write a CSV file with all the information (suffix <code>final.csv</code>) in the directory <code>snv</code>.</p>

<h4 id='amplicon_mode'>Amplicon mode</h4>

<p><code>amplian.py</code> will analyse a single window, corresponding to the reference sequence passed with the option <code>-f</code>. Consequently, the output files from the support to the SNV file will refer to this single window (see shotgun mode above).</p>
<hr />
<p>Go back <a href='index.html'>home</a></p>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">ShoRAH maintained by <a href="https://github.com/ozagordi">ozagordi</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
